content/blog/2021/06/writing-your-first-policy-with-kubewarden.md

```diff
@@ -0,0 +1,553 @@
+error:
+
+```console
+$ kwctl run policy.wasm --request-path request.json 2> /dev/null | jq
+{
+  "uid": "150d1761-43fb-4f36-bb73-7a3888a0bca2",
+  "allowed": false,
+  "status": {
+    "message": "namespace test-policy-ns does not comply with the company namespace naming convention: check https://internal.my-company.com/k8s/naming-conventions.md for more information"
+  }
+}
+```
+
+Now let's implement the rest of the checks:
+
+```go
+package main
+
+import (
+	"fmt"
+	"github.com/kubewarden/gjson"
+	kubewarden "github.com/kubewarden/policy-sdk-go"
+	"regexp"
+	wapc "github.com/wapc/wapc-guest-tinygo"
+)
+
+func main() {
+	wapc.RegisterFunctions(wapc.Functions{
+		"validate": func(payload []byte) ([]byte, error) {
+			namespaceName := gjson.Get(string(payload), "request.object.metadata.name")
+			namespaceNameMatches, err := regexp.MatchString(`^t-\w+-\w+`, namespaceName.String())
+			if err != nil {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("[compliance] internal error while processing namespace: %v", err)), kubewarden.NoCode)
+			}
+			if !namespaceNameMatches {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("[compliance] namespace %s does not comply with the company namespace naming convention: check https://internal.my-company.com/k8s/naming-conventions.md for more information", namespaceName)), kubewarden.NoCode)
+			}
+			teamContactEmail := gjson.Get(string(payload), "request.object.metadata.annotations.compliance\\.my-company\\.com/team-contact-email").String()
+			if teamContactEmail == "" {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("[compliance] the namespace %s is missing a mandatory annotation compliance.my-company.com/team-contact-email", namespaceName)), kubewarden.NoCode)
+			}
+			teamContactEmailValidDomain, err := regexp.MatchString(`@my-company\.com$`, teamContactEmail)
+			if err != nil {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("[compliance] internal error while processing namespace: %v", err)), kubewarden.NoCode)
+			}
+			if !teamContactEmailValidDomain {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("[compliance] the namespace %s does not comply with the company namespace naming convention: the compliance.my-company.com/team-contact-email annotation must end with @my-company.com", namespaceName)), kubewarden.NoCode)
+			}
+			teamRegion := gjson.Get(string(payload), "request.object.metadata.annotations.compliance\\.my-company\\.com/team-region").String()
+			if teamRegion != "AMER" && teamRegion != "APAC" && teamRegion != "EMEA" {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("[compliance] the namespace %s does not comply with the company namespace naming convention: the compliance.my-company.com/team-region annotation must be one of: AMER, APAC or EMEA; is: '%s'", namespaceName, teamRegion)), kubewarden.NoCode)
+			}
+			return kubewarden.AcceptRequest()
```

Comment:
Some comments:

* you are fetching multiple values from the json document. I would consider using `GetMany`. If you prefer to keep the code more understandable, you should use `GetByte`
* Do not write a regexp to check the email address: use instead an existing golang package like [this one](https://github.com/badoux/checkmail) I found with a quick Google search. The main reason for that: show you can build on existing libraries, without going crazy like with other solutions

Reply:
I tried checkmail but it wouldn't work with tinygo. In any case, this is a sample for an introductory blog post and I think it's good enough for the purpose.

Reply:
Given that `GetMany` returns a `[]Result` I think the code is more clear in this case with explicit `Get`.